File: app/src/main/java/com/omarea/krscript/WebViewInjector.java
package com.omarea.krscript;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.webkit.JavascriptInterface;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;

import com.omarea.common.shell.KeepShellPublic;
import com.omarea.common.shell.ShellExecutor;
import com.omarea.common.ui.DialogHelper;
import com.omarea.krscript.downloader.Downloader;
import com.omarea.krscript.executor.ExtractAssets;
import com.omarea.krscript.executor.ScriptEnvironmen;
import com.omarea.krscript.model.NodeInfoBase;
import com.omarea.krscript.model.ShellHandlerBase;
import com.omarea.krscript.ui.ParamsFileChooserRender;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Iterator;
import java.util.UUID;
import com.tool.tree.R;

public class WebViewInjector {
    private final WebView webView;
    private final Context context;
    private final ParamsFileChooserRender.FileChooserInterface fileChooser;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    @SuppressLint("SetJavaScriptEnabled")
    public WebViewInjector(WebView webView, ParamsFileChooserRender.FileChooserInterface fileChooser) {
        this.webView = webView;
        this.context = webView.getContext();
        this.fileChooser = fileChooser;
    }

    @SuppressLint({"JavascriptInterface", "SetJavaScriptEnabled"})
    public void inject(final Activity activity, final boolean credible) {
        if (webView != null) {
            WebSettings webSettings = webView.getSettings();
            
            // --- TỐI ƯU HÓA WEBVIEW ---
            webSettings.setJavaScriptEnabled(true);
            webSettings.setDomStorageEnabled(true); // Quan trọng để web chạy nhanh
            webSettings.setDatabaseEnabled(true);
            webSettings.setCacheMode(WebSettings.LOAD_DEFAULT); // Sử dụng cache hệ thống
            
            // Tạm thời chặn ảnh để ưu tiên tải cấu trúc HTML và Script Shell
            webSettings.setBlockNetworkImage(true);
            
            webSettings.setAllowFileAccess(credible);
            webSettings.setAllowUniversalAccessFromFileURLs(credible);
            webSettings.setAllowFileAccessFromFileURLs(credible);
            webSettings.setAllowContentAccess(true);
            webSettings.setUseWideViewPort(true);
            webSettings.setLoadWithOverviewMode(true);
            // ---------------------------

            webView.addJavascriptInterface(
                    new KrScriptEngine(context),
                    "KrScriptCore"
            );
            
            webView.setDownloadListener((url, userAgent, contentDisposition, mimetype, contentLength) -> {
                DialogHelper.Companion.animDialog(new AlertDialog.Builder(context)
                        .setTitle(R.string.kr_download_confirm)
                        .setMessage(url + "\n\n" + mimetype + "\n" + contentLength + "Bytes")
                        .setPositiveButton(R.string.btn_confirm, (dialog, which) -> 
                            new Downloader(context, null).downloadBySystem(url, contentDisposition, mimetype, UUID.randomUUID().toString(), null))
                        .setNegativeButton(R.string.btn_cancel, (dialog, which) -> {
                        })).setCancelable(false);
            });
        }
    }

    private class KrScriptEngine {
        private final Context context;
        private final NodeInfoBase virtualRootNode = new NodeInfoBase("");

        private KrScriptEngine(Context context) {
            this.context = context;
        }

        @JavascriptInterface
        public boolean rootCheck() {
            return KeepShellPublic.INSTANCE.checkRoot();
        }

        @JavascriptInterface
        public String executeShell(String script) {
            if (script != null && !script.isEmpty()) {
                return ScriptEnvironmen.executeResultRoot(context, script, virtualRootNode);
            }
            return "";
        }

        @JavascriptInterface
        public boolean executeShellAsync(String script, String callbackFunction, String env) {
            HashMap<String, String> params = new HashMap<>();
            Process process = null;
            try {
                if (env != null && !env.isEmpty()) {
                    JSONObject paramsObject = new JSONObject(env);
                    for (Iterator<String> it = paramsObject.keys(); it.hasNext(); ) {
                        String key = it.next();
                        params.put(key, paramsObject.getString(key));
                    }
                }
                process = ShellExecutor.getSuperUserRuntime();
            } catch (Exception ex) {
                Toast.makeText(context, ex.getMessage(), Toast.LENGTH_SHORT).show();
            }

            if (process != null) {
                final OutputStream outputStream = process.getOutputStream();
                final DataOutputStream dataOutputStream = new DataOutputStream(outputStream);

                setHandler(process, callbackFunction, () -> { });

                ScriptEnvironmen.executeShell(context, dataOutputStream, script, params, null, null);
                return true;
            } else {
                return false;
            }
        }

        @JavascriptInterface
        public String extractAssets(String assets) {
            return new ExtractAssets(context).extractResource(assets);
        }

        @JavascriptInterface
        public boolean fileChooser(final String callbackFunction) {
            if (fileChooser != null) {
                return fileChooser.openFileChooser(new ParamsFileChooserRender.FileSelectedInterface() {
                    @Override
                    public int type() {
                        return ParamsFileChooserRender.FileSelectedInterface.Companion.getTYPE_FILE();
                    }

                    @Nullable
                    @Override
                    public String suffix() {
                        return null;
                    }

                    @NotNull
                    @Override
                    public String mimeType() {
                        return "*/*";
                    }

                    @Override
                    public void onFileSelected(@Nullable String path) {
                        // Trả dữ liệu về UI thread để tránh crash khi gọi JS
                        mainHandler.post(() -> {
                            try {
                                final JSONObject message = new JSONObject();
                                message.put("absPath", (path == null || path.isEmpty()) ? null : path);
                                webView.evaluateJavascript(callbackFunction + "(" + message + ")", null);
                            } catch (Exception ignored) { }
                        });
                    }
                });
            }
            return false;
        }

        private void setHandler(Process process, final String callbackFunction, final Runnable onExit) {
            final InputStream inputStream = process.getInputStream();
            final InputStream errorStream = process.getErrorStream();
            
            // Thread đọc luồng Standard Output
            final Thread reader = new Thread(() -> {
                try (BufferedReader br = new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8))) {
                    String line;
                    while ((line = br.readLine()) != null) {
                        sendJsLog(callbackFunction, ShellHandlerBase.EVENT_REDE, line + "\n");
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            });

            // Thread đọc luồng Error Output
            final Thread readerError = new Thread(() -> {
                try (BufferedReader br = new BufferedReader(new InputStreamReader(errorStream, StandardCharsets.UTF_8))) {
                    String line;
                    while ((line = br.readLine()) != null) {
                        sendJsLog(callbackFunction, ShellHandlerBase.EVENT_READ_ERROR, line + "\n");
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            });

            final Process processFinal = process;
            Thread waitExit = new Thread(() -> {
                int status = -1;
                try {
                    status = processFinal.waitFor();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                } finally {
                    sendJsLog(callbackFunction, ShellHandlerBase.EVENT_EXIT, String.valueOf(status));

                    if (reader.isAlive()) reader.interrupt();
                    if (readerError.isAlive()) readerError.interrupt();
                    if (onExit != null) onExit.run();
                }
            });

            reader.start();
            readerError.start();
            waitExit.start();
        }

        // Hàm hỗ trợ gửi log về WebView thông qua mainHandler để ổn định hiệu suất
        private void sendJsLog(String callback, int type, String messageStr) {
            mainHandler.post(() -> {
                try {
                    final JSONObject message = new JSONObject();
                    message.put("type", type);
                    message.put("message", messageStr);
                    webView.evaluateJavascript(callback + "(" + message + ")", null);
                } catch (Exception ignored) { }
            });
        }
    }
}

File: app/src/main/kotlin/com/omarea/krscript/WebViewInjector.kt
package com.omarea.krscript

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.webkit.JavascriptInterface
import android.webkit.WebSettings
import android.webkit.WebView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import com.omarea.common.shell.KeepShellPublic
import com.omarea.common.shell.ShellExecutor
import com.omarea.common.ui.DialogHelper
import com.omarea.common.ui.ProgressBarDialog
import com.omarea.krscript.downloader.Downloader
import com.omarea.krscript.executor.ExtractAssets
import com.omarea.krscript.executor.ScriptEnvironmen
import com.omarea.krscript.model.NodeInfoBase
import com.omarea.krscript.model.ShellHandlerBase
import com.omarea.krscript.ui.ParamsFileChooserRender
import com.tool.tree.R
import org.json.JSONObject
import java.io.BufferedReader
import java.io.DataOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.util.UUID

class WebViewInjector(
    private val webView: WebView,
    private val fileChooser: ParamsFileChooserRender.FileChooserInterface?
) {
    private val context: Context = webView.context
    private val mainHandler = Handler(Looper.getMainLooper())

    // Dialog loading dùng chung layout dialog_loading.xml, có sẵn nút Hủy (dialog_cancel_button)
    // để người dùng có thể hủy việc tải trang khi đang duyệt web.
    private var loadingDialog: ProgressBarDialog? = null

    @SuppressLint("JavascriptInterface", "SetJavaScriptEnabled")
    fun inject(activity: Activity, credible: Boolean) {
        loadingDialog = ProgressBarDialog(activity, null)

        val webSettings: WebSettings = webView.settings

        // --- TỐI ƯU HÓA WEBVIEW ---
        webSettings.javaScriptEnabled = true
        webSettings.domStorageEnabled = true // Quan trọng để web chạy nhanh
        webSettings.databaseEnabled = true
        webSettings.cacheMode = WebSettings.LOAD_DEFAULT // Sử dụng cache hệ thống

        // Tạm thời chặn ảnh để ưu tiên tải cấu trúc HTML và Script Shell
        webSettings.blockNetworkImage = false

        webSettings.allowFileAccess = credible
        webSettings.allowUniversalAccessFromFileURLs = credible
        webSettings.allowFileAccessFromFileURLs = credible
        webSettings.allowContentAccess = true
        webSettings.useWideViewPort = true
        webSettings.loadWithOverviewMode = true
        // ---------------------------

        webView.addJavascriptInterface(
            KrScriptEngine(context),
            "KrScriptCore"
        )

        webView.setDownloadListener { url, _, contentDisposition, mimetype, contentLength ->
            DialogHelper.Companion.animDialog(
                AlertDialog.Builder(context)
                    .setTitle(R.string.kr_download_confirm)
                    .setMessage("$url\n\n$mimetype\n${contentLength}Bytes")
                    .setPositiveButton(R.string.btn_confirm) { _, _ ->
                        Downloader(context, null).downloadBySystem(
                            url, contentDisposition, mimetype, UUID.randomUUID().toString(), null
                        )
                    }
                    .setNegativeButton(R.string.btn_cancel) { _, _ -> }
            ).setCancelable(false)
        }
    }

    /**
     * Hiển thị dialog loading (layout dialog_loading.xml) kèm nút Hủy khi đang tải trang web.
     * Bấm nút Hủy sẽ gọi webView.stopLoading() để dừng việc tải trang giữa chừng.
     * Gọi hàm này trong WebViewClient.onPageStarted khi duyệt web.
     */
    fun showLoading(message: String) {
        val dialog = loadingDialog ?: return
        dialog.showDialogWithCancel(message) {
            webView.stopLoading()
            kotlin.Unit
        }
    }

    fun showLoading() {
        showLoading(context.getString(R.string.please_wait))
    }

    /**
     * Ẩn dialog loading. Gọi trong WebViewClient.onPageFinished (hoặc khi tải lỗi)
     * khi duyệt web.
     */
    fun hideLoading() {
        loadingDialog?.hideDialog()
    }

    private inner class KrScriptEngine(private val context: Context) {
        private val virtualRootNode = NodeInfoBase("")

        @JavascriptInterface
        fun rootCheck(): Boolean {
            return KeepShellPublic.checkRoot()
        }

        @JavascriptInterface
        fun executeShell(script: String?): String {
            if (!script.isNullOrEmpty()) {
                return ScriptEnvironmen.executeResultRoot(context, script, virtualRootNode)
            }
            return ""
        }

        @JavascriptInterface
        fun executeShellAsync(script: String?, callbackFunction: String, env: String?): Boolean {
            val params = HashMap<String, String>()
            var process: Process? = null
            try {
                if (!env.isNullOrEmpty()) {
                    val paramsObject = JSONObject(env)
                    val it = paramsObject.keys()
                    while (it.hasNext()) {
                        val key = it.next()
                        params[key] = paramsObject.getString(key)
                    }
                }
                process = ShellExecutor.getSuperUserRuntime()
            } catch (ex: Exception) {
                Toast.makeText(context, ex.message, Toast.LENGTH_SHORT).show()
            }

            return if (process != null) {
                val outputStream = process.outputStream
                val dataOutputStream = DataOutputStream(outputStream)

                setHandler(process, callbackFunction) { }

                ScriptEnvironmen.executeShell(context, dataOutputStream, script, params, null, null)
                true
            } else {
                false
            }
        }

        @JavascriptInterface
        fun extractAssets(assets: String): String {
            return ExtractAssets(context).extractResource(assets)
        }

        @JavascriptInterface
        fun fileChooser(callbackFunction: String): Boolean {
            val chooser = fileChooser ?: return false
            return chooser.openFileChooser(object : ParamsFileChooserRender.FileSelectedInterface {
                override fun type(): Int {
                    return ParamsFileChooserRender.FileSelectedInterface.Companion.TYPE_FILE
                }

                override fun suffix(): String? {
                    return null
                }

                override fun mimeType(): String {
                    return "*/*"
                }

                override fun onFileSelected(path: String?) {
                    // Trả dữ liệu về UI thread để tránh crash khi gọi JS
                    mainHandler.post {
                        try {
                            val message = JSONObject()
                            message.put("absPath", if (path.isNullOrEmpty()) null else path)
                            webView.evaluateJavascript("$callbackFunction($message)", null)
                        } catch (ignored: Exception) {
                        }
                    }
                }
            })
        }

        private fun setHandler(process: Process, callbackFunction: String, onExit: Runnable) {
            val inputStream: InputStream = process.inputStream
            val errorStream: InputStream = process.errorStream

            // Thread đọc luồng Standard Output
            val reader = Thread {
                try {
                    BufferedReader(InputStreamReader(inputStream, StandardCharsets.UTF_8)).use { br ->
                        var line: String?
                        while (br.readLine().also { line = it } != null) {
                            sendJsLog(callbackFunction, ShellHandlerBase.EVENT_REDE, line + "\n")
                        }
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }

            // Thread đọc luồng Error Output
            val readerError = Thread {
                try {
                    BufferedReader(InputStreamReader(errorStream, StandardCharsets.UTF_8)).use { br ->
                        var line: String?
                        while (br.readLine().also { line = it } != null) {
                            sendJsLog(callbackFunction, ShellHandlerBase.EVENT_READ_ERROR, line + "\n")
                        }
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }

            val processFinal = process
            val waitExit = Thread {
                var status = -1
                try {
                    status = processFinal.waitFor()
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                } finally {
                    sendJsLog(callbackFunction, ShellHandlerBase.EVENT_EXIT, status.toString())

                    if (reader.isAlive) reader.interrupt()
                    if (readerError.isAlive) readerError.interrupt()
                    onExit.run()
                }
            }

            reader.start()
            readerError.start()
            waitExit.start()
        }

        // Hàm hỗ trợ gửi log về WebView thông qua mainHandler để ổn định hiệu suất
        private fun sendJsLog(callback: String, type: Int, messageStr: String) {
            mainHandler.post {
                try {
                    val message = JSONObject()
                    message.put("type", type)
                    message.put("message", messageStr)
                    webView.evaluateJavascript("$callback($message)", null)
                } catch (ignored: Exception) {
                }
            }
        }
    }
}
